cpp/service_layer_service_impl.cc

```diff
@@ -0,0 +1,75 @@
+#include "service_layer_service_impl.h"
+
+grpc::Status ServiceLayerServiceImpl::registeruser(grpc::ServerContext* context, const chirp::RegisterRequest* request, chirp::RegisterReply* reply) {
+  std::string username = request->username();
+  bool response = service_.registeruser(username);
+  if (!response) {
+    grpc::Status existing_user = grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Usernames must be unique");
+    return existing_user;
+  }
+  return grpc::Status::OK;
+}
+
+grpc::Status ServiceLayerServiceImpl::chirp(grpc::ServerContext* context, const chirp::ChirpRequest* request, chirp::ChirpReply* reply) {
+  chirp::Chirp this_chirp = service_.chirp(request->username(), request->text(), request->parent_id());
+  if (this_chirp.id() == "ERROR") {
+    grpc::Status parent_not_found = grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Invalid Parent ID");
+    return parent_not_found;
+  }
+  chirp::Chirp* reply_chirp = reply->mutable_chirp();
+  *reply_chirp = this_chirp;
+  return grpc::Status::OK;
+}
+grpc::Status ServiceLayerServiceImpl::follow(grpc::ServerContext* context, const chirp::FollowRequest* request, chirp::FollowReply* reply) {
+  bool success = service_.follow(request->username(), request->to_follow());
+  if (!success) {
+    grpc::Status follow_failed = grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Follow failed");
+    return follow_failed;
+  }
+  return grpc::Status::OK;
+}
+
+grpc::Status ServiceLayerServiceImpl::read(grpc::ServerContext* context, const chirp::ReadRequest* request, chirp::ReadReply* reply) {
+  std::deque<chirp::Chirp> read_chirps = service_.read(request->chirp_id());
+  for (chirp::Chirp c : read_chirps) {
+    chirp::Chirp* chirp_pointer = reply->add_chirps();
+    const chirp::Chirp& added_chirp = c;
```

Comment:
const should start with a k and also these chirps can be declared outside of the loop and then filled in once you enter the loop